hibernate-core/src/main/java/org/hibernate/dialect/MySQLDialect.java

```diff
@@ -534,6 +534,33 @@ public JdbcType resolveSqlTypeDescriptor(
 					jdbcTypeCode = GEOMETRY;
 				}
 				break;
+			case Types.TINYINT:
+				if ( isUnsigned( columnTypeName ) ) {
+					return jdbcTypeRegistry.getDescriptor( Types.SMALLINT );
+				}
+				break;
+			case Types.SMALLINT:
+				if ( isUnsigned( columnTypeName ) ) {
+					return columnTypeName.contains( "TINYINT" ) ? jdbcTypeRegistry.getDescriptor( Types.SMALLINT )
+							: jdbcTypeRegistry.getDescriptor( Types.INTEGER );
+				}
+				break;
+			case Types.INTEGER:
+				if ( isUnsigned( columnTypeName ) ) {
+					if ( columnTypeName.contains( "SMALLINT" ) ) {
+						return jdbcTypeRegistry.getDescriptor( Types.INTEGER );
+					}
+					else if ( !columnTypeName.contains( "MEDIUMINT" ) ) {
+						return jdbcTypeRegistry.getDescriptor( Types.BIGINT );
+					}
+				}
+				break;
+			case Types.BIGINT:
+				if ( isUnsigned( columnTypeName ) ) {
+					return columnTypeName.contains( "INTEGER" ) ? jdbcTypeRegistry.getDescriptor( Types.BIGINT )
+							: jdbcTypeRegistry.getDescriptor( Types.NUMERIC );
+				}
+				break;
```

Comment:
Woah. This is ... a bit ugly. Surely there must be some more elegant way to do this?

Why not just directly switch on the `columnTypeName`, since you're depending on it anyway?